bits_test.go

```diff
@@ -196,3 +196,51 @@ func sortedUnique(nums []uint32) []uint32 {
 	}
 	return filtered
 }
+
+func TestCondenseRuneOffsets(t *testing.T) {
+	for i, tc := range []struct {
+		arr  []uint32
+		want []runeOffset
+	}{
+		{[]uint32{}, []runeOffset{}},
+		{[]uint32{0, 100, 200, 300, 400, 500}, []runeOffset{}},
+		{[]uint32{0, 105, 205, 310, 420}, []runeOffset{{100, 105}, {300, 310}, {400, 420}}},
+	} {
+		got := makeRuneOffsetMap(tc.arr)
+		if !reflect.DeepEqual(got, runeOffsetMap(tc.want)) {
+			t.Errorf("#%d: got %v, want %v", i, got, tc.want)
+		}
+	}
+}
+
+func TestRuneOffsetLookup(t *testing.T) {
+	for i, tc := range []struct {
+		r                 uint32
+		wantOff, wantLeft uint32
+		offsets           []runeOffset
+	}{
+		{0, 0, 0, nil},
+		{1234, 1200, 34, nil},
+		{5, 0, 5, []runeOffset{{100, 105}, {400, 430}}},
+		{120, 105, 20, []runeOffset{{100, 105}, {400, 430}}},
+		{1234, 1230, 34, []runeOffset{{100, 105}, {400, 430}}},
+	} {
+		gotOff, gotLeft := runeOffsetMap(tc.offsets).lookup(tc.r)
+		if gotLeft != tc.wantLeft {
+			t.Errorf("#%d: got left=%v, want left=%v", i, gotLeft, tc.wantLeft)
+		}
+		if gotOff != tc.wantOff {
+			t.Errorf("#%d: got off=%v, want off=%v", i, gotOff, tc.wantOff)
+		}
+	}
+
+	m := runeOffsetMap([]runeOffset{{100, 105}, {200, 210}, {400, 430}})
+	inputs := []uint32{0, 1, 99, 100, 101, 199, 200, 201, 300, 399, 400, 401, 510, 610}
+	wanted := []uint32{0, 0, 0, 105, 105, 105, 210, 210, 310, 310, 430, 430, 530, 630}
```

Comment:
I think the exhaustive test is still useful since its hard as a reader of the code to know your two arrays here are correct. You have two implementations here: the new one using runeOffsetMap and the old one just using runeOffsets. It seems useful to compare they agree on all possible inputs for a smallish range.

To make it a little more clear my idea: you take a `runeOffset []uint32` then compare the old algorithm (just dividing by frequency and looking up) and the new one via the map.